### ISSUE TITLE
Analyze design point number vs downscaling model skill

### ISSUE THREAD

**Issue by dlebauer** [ccmmf/organization#192]
- Generate 1000 design points
- Run ensembles at each design point
- For some set of n, e.g. 50, 100, 200, 500, 1000
  - choose representative subset of sites. these should not be full random, but based on a selection of sites based on distance from cluster center. This is how we choose design points from clusters in practice. 
  - fit downscaling model
  - downscale to all CA fields
  - record downscaling error [run multiple tests for n < 1000]
- Plot error as a function of n
- Evaluate if it follows a clear pattern;  we have sufficient coverage

This should be a reproducible analysis that can be repeated later.

Compare to other sources of uncertainty.

**Comment by divine7022** [comment]
Dropping this here for the record: the learning curve study of downscaling skill as a function of the number of design points, for the 1,000 point run we currently have.

## Method

For a given set of design points, fit a random forest (ranger, 1,000 trees) on a random 75% of them and score R2 and RMSE on the held out 25%. Predictors are the 28 downscaling covariates used in `040`. Script: `023_skill_vs_n.R`.

At each anchor size n in {100, 200, 350, 520, 750, 1000}:

- draw 25 random subsets of the truth points (the box spread), and
- take the farthest-point-sampling nested prefix of size n (the space filling line),

each scored per ensemble member and per variable, then summarized by median and interquartile range across subsets and members. Variables: AGB, NPP, TotSoilCarb. This is the single 1,000 point sequence for n <= 1000, which is the truth we have; n > 1000 waits on the 10k scale-up.(got runs for 1000, but don't have a runs for 10k yet)

## Results

Held out R2, combined PFTs, median across subsets and ensemble members:

| variable | method | n=100 | n=200 | n=350 | n=520 | n=750 | n=1000 |
| --- | --- | --- | --- | --- | --- | --- | --- |
| AGB | FPS | 0.14 | 0.34 | 0.30 | 0.34 | 0.31 | 0.33 |
| AGB | random | 0.11 | 0.21 | 0.28 | 0.30 | 0.33 | 0.35 |
| NPP | FPS | -0.06 | 0.19 | 0.36 | 0.29 | 0.33 | 0.35 |
| NPP | random | 0.07 | 0.19 | 0.28 | 0.31 | 0.36 | 0.38 |
| TotSoilCarb | FPS | 0.41 | 0.50 | 0.50 | 0.54 | 0.50 | 0.49 |
| TotSoilCarb | random | 0.22 | 0.35 | 0.41 | 0.45 | 0.47 | 0.49 |

What this shows:

- Space filling (FPS) beats random selection at small n, and the two converge by n around 750 to 1000. Clearest for TotSoilCarb: FPS reaches R2 0.41 at n=100 against 0.22 for random, and both sit near 0.49 at n=1000.
- Skill plateaus early. TotSoilCarb is at roughly 0.50 by n around 200 and does not improve past it. AGB and NPP plateau lower, around 0.33 to 0.38. Past a few hundred points, adding design points does not raise held out R2 in this study; the ceiling looks set by the covariates, not the sample size.
- NPP RMSE is on the order of 5e-9 because NPP is carried in per-second flux units, so RMSE is not comparable across variables here. R2 is the metric to read.
- The `e(n) = a + b * n^-c` fit (`design_point_skill_fit.csv`) did not converge to a usable extrapolation. The RMSE curves flatten by n around 200 and are not cleanly power-law (the fitted exponent is degenerate for AGB and TotSoilCarb), so the plateau is read empirically rather than off the fit.

## Figures

For each variable, three panels (x axis is number of design points):

- `skill_vs_n_totsoilcarb.svg`, `skill_vs_n_agb.svg`, `skill_vs_n_npp.svg` -- CV R2 vs n; boxes are the random subsets, the blue line is the FPS space-filling design

`skill_vs_n_totsoilcarb.svg`:

<img width="768" height="432" alt="Image" src="https://github.com/user-attachments/assets/49d27a79-2310-428e-ae4e-d60935921631" />

`skill_vs_n_agb.svg`:

<img width="768" height="432" alt="Image" src="https://github.com/user-attachments/assets/c066fd91-71b5-4a06-8144-2a3e74b2b793" />

`skill_vs_n_npp.svg`:

<img width="768" height="432" alt="Image" src="https://github.com/user-attachments/assets/58475e0b-da10-45df-ab73-52ec69f6b3f4" />

- `skill_vs_n_totsoilcarb_r2.svg`, `skill_vs_n_agb_r2.svg`, `skill_vs_n_npp_r2.svg` -- CV R2 vs n

`skill_vs_n_totsoilcarb_r2.svg`:

<img width="768" height="432" alt="Image" src="https://github.com/user-attachments/assets/6f2844d2-d96a-4c20-8430-6f35b767089f" />

`skill_vs_n_agb_r2.svg`:

<img width="768" height="432" alt="Image" src="https://github.com/user-attachments/assets/bd207eb4-bb7a-4659-9f67-df6a40e18e5f" />

`skill_vs_n_npp_r2.svg`:

<img width="768" height="432" alt="Image" src="https://github.com/user-attachments/assets/71f4b9b8-8914-438b-9581-7425a0fbd5a9" />


- `skill_vs_n_totsoilcarb_rmse.svg`, `skill_vs_n_agb_rmse.svg`, `skill_vs_n_npp_rmse.svg` -- CV RMSE vs n

`skill_vs_n_totsoilcarb_rmse.svg`:

<img width="768" height="432" alt="Image" src="https://github.com/user-attachments/assets/d689b0ee-4299-42eb-b8f3-8b04e1bfde1e" />

`skill_vs_n_agb_rmse.svg`:

<img width="768" height="432" alt="Image" src="https://github.com/user-attachments/assets/002d8e09-421f-4c3c-b4b5-d0d3fbaafe54" />

`skill_vs_n_npp_rmse.svg`:

<img width="768" height="432" alt="Image" src="https://github.com/user-attachments/assets/20877937-8f8c-4332-bf53-52bee51bea3d" />


## Caveats

- The random 75/25 holdout measures the random forest's interpolation skill over covariate space at that sample size. It is not spatial extrapolation to held-out regions; a spatially blocked holdout gives a lower number and answers a different question.
- Single truth realization (the 1,000-point sipnet run). There is no separate n > 1000 sipnet runs yet.
- Outputs: `design_point_skill_{curve,summary,fit}.csv`.

**Comment by dlebauer** [comment]
This is a nice analysis, thank you for pulling it together.

To clarify, the goal is to answer the question: Find _n_ such that replacing site-level SIPNET runs with RF predictions does not meaningfully increase uncertainty. Specifically, the RF approximation MSE should be less than 1% of the corresponding ensemble variance. If 1% is not practical, we could set the threshold to 10%. Bias should also be reported.

As you point out, the plateau implies that increasing _n_ will be insufficient to reduce approximation error, but the current analysis does not isolate covariates as the only cause. 
The plateau in RMSE and R2 could be due to
- insufficient covariates or covariate transformations - the climate variables (MAT, MAP, VPD) were only intended as a first pass; it is worth replacing these with BioClim variables
- insufficient model - e.g. we could try cnn or gradient boosting
- a bug in the code or other error 

Another source of error is that the 75/25 split means that each skill metric is calculated on a different set of validation points. It should be possible to remove this error by using a single, fixed set of validation sites for all outputs and values of n.

Moving forward, lets prioritize the following metrics (but include others deemed useful):
- IQR Normalized RMSE for comparing across outputs
- R2 as a secondary diagnostic
- Bias

You can stick with FPS - it is not necessary to further compare it to the random sampling approach.

**Comment by divine7022** [comment]
Reran the learning curve with the changes you asked for. Result in short: for TotSoilCarb and NPP
the curve flattens well short of the tolerance and the limit is not sample size. For AGB I cannot
draw that conclusion, because the AGB truth from this run is not physical and the curve is still
descending. A rerun is needed either way, and I have scoped it below.

## What changed since pass 1

- **One fixed validation set** (#264). 202 sites drawn once, stratified by k-means (k=10) over the
  scaled covariate space within each PFT, so the holdout represents covariate space rather than an
  unlucky draw. The same 202 sites score every output and every `n`. Training pool is the other 798.
- `randomForest`, `ntree = 1000`, 80/20 ratio, and the same nodesize rule
  as `ensemble_downscale()`. Two deliberate differences: I fit one RF per PFT where production fits a
  single pooled RF with the PFT indicators as live predictors, and I predict only the 202 held-out
  design points rather than the 550k parcels.
- **Uncertainty from repeating the design.** Five FPS orderings from different start points at each
  `n`. These are correlated by construction, not independent samples.
- **Acceptance metric is surrogate error against ensemble variance**, with IQR-normalized RMSE, R2
  and bias alongside. FPS only; dropped the random-sampling comparison.
.

## Results

RMSE / ensemble SD, combined PFT, median across 5 designs x 20 members. Target 0.20; the
1% of variance bar is 0.10.

| variable | n=100 | n=200 | n=350 | n=500 | n=650 | n=800 | last step |
| --- | --- | --- | --- | --- | --- | --- | --- |
| TotSoilCarb | 1.23 | 1.21 | 1.16 | 1.14 | 1.13 | 1.14 | +0.005 |
| AGB | 0.97 | 0.94 | 0.89 | 0.88 | 0.85 | 0.82 | -0.029 |
| NPP | 0.86 | 0.81 | 0.79 | 0.77 | 0.75 | 0.75 | -0.006 |

**TotSoilCarb and NPP have flattened.** TotSoilCarb's last step is a slight increase; NPP's is
-0.006 against -0.053 on the first step. Fitting `e(n) = a + b n^-c` gives a stable asymptote of
**0.945 (TotSoilCarb)** and **0.609 (NPP)**, both far above 0.20, so no achievable `n` clears the
bar for those two. TotSoilCarb needs a 5.7x error reduction.

**AGB has not flattened.** Its last step (-0.029) is the second largest of the five, and R2 is still
climbing (0.20 at n=100 to 0.42 at n=800). The power-law fit is unstable for AGB: of four start
values only one converges, and it returns a negative exponent, meaning error rising with `n`, which
is not usable. The other three fail to converge. So I have no defensible extrapolation for AGB, and
I am not claiming more design points would fail to help it.

Two caveats on the design itself:

- **`n = 800` is the entire training pool** (416 annual + 382 woody = 798), so all five FPS orderings
  collapse to the same training set there. The spread across designs falls from about 1-2% at n=100
  to 0.2-0.5% at n=800. The last point on the curve carries no design variability.
- The GHG outputs are absent from this table for the reason in the next section.

## The AGB truth from this run is not physical

`AGB = plantWoodC + plantLeafC` at the last month is **negative for 51.5% of site-member cells**
(annual crop 63.2%, woody perennial 38.8%), with a minimum of **-43.4 kg C m-2** and 24.2% of cells
below -0.01. The annual-crop total across the validation month sums to **-273 kg C m-2**.

I previously described this as "near zero at the seasonal minimum," quoting the median (-4.4e-05).
That was the one number that made it look benign, and it was the wrong way to report it. This is a
mass-balance problem in the truth data, not a small-value artifact, and it undermines the AGB skill
numbers independently of anything about design points.

## Diagnostic on the plateau

At n=800 I compared out-of-bag error to held-out test error. This was a separate diagnostic: a
single pooled RF over both PFTs, `nodesize = 5`, averaged over members 1-5, so it is not the same
fit that produced the curve.

| variable | OOB RMSE | test RMSE | test / OOB |
| --- | --- | --- | --- |
| TotSoilCarb | 2.93 | 2.86 | 0.98 |
| AGB | 0.774 | 0.661 | 0.85 |
| NPP | 6.23e-9 | 5.74e-9 | 0.92 |

Test error is close to OOB on average, which is consistent with a covariate or model ceiling rather
than overfitting. I would not push it further than that. RF OOB predictions average only about a
third of the trees and are therefore pessimistic relative to a full-forest test prediction, so a
ratio in 0.85-0.98 is roughly what that bias alone produces. Per member the picture is not uniform:
4 of 15 cells have test above OOB. So this is suggestive, not decisive.

The clearer evidence for a ceiling is the shape of the TotSoilCarb and NPP curves and their fitted
asymptotes, not this ratio.

## Bias

Bias relative to ensemble SD is small across all rows (0.003 to 0.166).

Aggregate relative bias needs care. Defined as `(sum(pred) - sum(actual)) / sum(actual)` over the
202 validation sites, it is only interpretable when the denominator is well away from zero. That
holds for **TotSoilCarb: 2 to 5 percent**, which is well behaved. It does not hold for AGB, where the
per-member sum ranges from -75 to +140 and the ratio spans -5.3 to +6.1 across cells; NPP is
intermediate. So I would report the TotSoilCarb aggregate bias and treat the AGB and NPP versions as
uninformative until the truth data is fixed.

## Two observations

- **R2 and the acceptance metric rank the outputs in opposite order.** At n=800, R2 is TotSoilCarb
  0.49 > AGB 0.42 > NPP 0.33, while the acceptance ratio is NPP 0.75 < AGB 0.82 < TotSoilCarb 1.14.
  The reason is the denominator: R2 divides by between-site variance, the acceptance metric by
  across-member variance, and on the validation sites the spatial SD exceeds the ensemble SD for all
  three outputs (ratios of about 1.2 to 1.5, depending on how the spatial term is pooled). So R2 is
  measured against the larger denominator throughout, and leading with it, as pass 1 did,
  understates the gap.
- `ocd` is the top TotSoilCarb predictor by a wide margin in the member I inspected (%IncMSE 42.3 vs
  28.9 for the next). For NPP, `area_ha` ranks first in 8 of 13 members I checked but as low as 17th
  in others, so I would not state that one as a general result.

## Why a rerun is required

Three separate reasons, all of which point at the same run:

1. **GHG outputs are unusable.** `N2O_flux` and `CH4_flux` are exactly zero in every row, because the run has flags `NITROGEN_CYCLE = 0` and `ANAEROBIC = 0`. This needs a rerun, not a re-extraction.
2. **The AGB truth is not physical** (above), and the SIPNET changes since our binary target exactly
   that.
3. The negative-growth change is the relevant one. The old code did not allocate negative mean NPP to
   pools; current code allocates it and adds `checkNegativeCreation()` and a mortality check.
  That is precisely the regime producing our negative AGB.

   I want to correct something I would otherwise have implied: I cannot claim TotSoilCarb is
   unaffected by the version change. #301 is the only change to the soil-C update equation and it is
   off, but `TotSoilCarb = soil + litter`, and #337 and #368 both alter the root and litter fluxes
   feeding those pools. The TotSoilCarb magnitudes are provisional too. What does not depend on the
   rerun is the shape of the curve and the size of the gap, which is 5.7x for TotSoilCarb.

   Separately, `leafOnReallocFrac` is registered as required, and our `sipnet.param` does not contain it,
   so current SIPNET will exit with an input-file error against our existing parameter files. The rerun needs 
   updated inputs, not just a rebuild.

## Proposed rerun scope

one run of the 1k design that covers all of it:

- current SIPNET at a pinned commit, with parameter files updated for `leafOnReallocFrac`
- `NITROGEN_CYCLE = 1` and `ANAEROBIC = 1` so the GHG outputs are usable
- evaluation target for the seasonal outputs chosen deliberately rather than defaulting to the last
  month, since December is the seasonal minimum for biomass and productivity
- extend the `n` ladder above 800 so the last anchor is not the whole training pool

Does that scope look right to you, and is there anything you want added while the run is going?

## On #213

I do not think `n` can be set on skill grounds from this pass. TotSoilCarb and NPP fail at every `n`
with flat curves, and AGB is unresolved pending a physical truth. On current evidence the choice of
1000 is a compute-budget decision rather than a skill-justified one, and I would rather say that than
pick a number the curve does not support.

## On #265

I will run the BioClim ablation regardless. One question: since the covariates set the ceiling here,
would you also want an arm using the model's own inputs as covariates? We are emulating a model, so
what bounds the RF is how well the covariates encode what SIPNET consumed -- met forcing, the
management events in `events.in`, and the pool initial conditions. For a slow pool over eight years
the end state is dominated by its initial condition, and `ocd` is the closest proxy in the stack,
which is consistent with it ranking first for TotSoilCarb.

To decide between those arms empirically rather than by argument, I would add an EOF decomposition of
the residual field over sites. If the residuals carry coherent spatial structure, that points at a missing spatially 
structured covariate and tells us which one to chase; if they are unstructured, the ceiling is closer to
irreducible at this support and the covariate search has a limit worth knowing about.

**Comment by dlebauer** [comment]
Thanks @divine7022 this is a very nice analysis. It is clear that the relevance of the model runs with incorrect carbon accounting, uncalibrated parameters, and no GHG.

I agree we should use the 1,000 design points and move forward with the uncertainty analysis. The current results are useful: they show that increasing the number of points alone is unlikely to solve the downscaling error, and the BioClim analysis suggests that this path is not promising.

The large grid size of met data is an issue that we can also defer.

We have enough information to support 1000 design points for now and defer further work until 

- model is calibrated 
- time and space varying events and pft parameters are used in both sipnet and RF inputs
- ideally, higher resolution meteorology 

Now I'm leaving some thoughts for when we return, summarizing many of your results and conclusions

1. Build covariates directly from the inputs used by SIPNET. Make sure the emulator has the information SIPNET receives, including:
    * initial conditions, especially soil C, LAI/biomass - but see 2 below 
    * spatially and temporally varying management events and crop/PFT parameters
    * appropriate summaries of the meteorological forcing (maybe bioclim vars will explain more if met drivers have more variance)
2. Predict **change** in soil and perhaps woody agb carbon pools as the downscaling targets and add initial SOC / AGB back afterward.
3. Re-evaluate the learning curve with these model-native predictors.
4. (Maybe worth evaluating) Test whether ensemble members can be modeled jointly. Instead of fitting a completely separate emulator for each ensemble member, test a pooled model with SIPNET parameter values and member-specific initial conditions as predictors. This could make better use of the available site x ensemble simulations. **But** should confirm that it it preserves SIPNET ensemble spread and covariance on held-out sites in order to ensure the independence of ensemble members 
5. Explore alternative ML models only after improving inputs.

I would defer variables that are not directly derived from what SIPNET sees (eg pH, sand, additional SoilGrids variables, different sources of inputs).

And to reiterate, thank you for all of your work on implementing, analyzing, and interpreting. It's hard work, but we are making progress!
